src/test/java/org/opensearch/knn/recall/RecallTestsIT.java

```diff
@@ -0,0 +1,183 @@
+        }
+        else
+            indexVectors = TestUtils.generateRandomVectors(docCount, dimensions);
+
+        bulkAddKnnDocs(testIndex, testField, indexVectors, docCount);
+    }
+
+    public void bulkAddKnnDocs(String index, String fieldName,  float[][] indexVectors, int docCount) throws IOException {
+        Request request = new Request(
+                "POST",
+                 "/_bulk"
+        );
+
+        request.addParameter("refresh", "true");
+        String body = "";
+
+        for(int i = 0; i < docCount; i++){
+          body += "{ \"index\" : { \"_index\" : \"" + index + "\", \"_id\" : \"" + String.valueOf(i+1) + "\" } }\n"
+          +"{ \"" + fieldName + "\" : "+ Arrays.toString(indexVectors[i]) +" }\n";
+        }
+
+        request.setJsonEntity(body);
+
+        Response response = client().performRequest(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 200);
+
+    }
+
+    public double getRecallValue(String testIndex, String testField, int docCount, int queryCount, int k, int dimensions, SpaceType spaceType, boolean isStandard) throws Exception{
+        float[][] indexVectors, queryVectors;
+
+        indexVectors = getIndexVectorsFromIndex(testIndex, testField, docCount, dimensions);
+
+        if(isStandard)
+            queryVectors = TestUtils.generateQueryVectors(queryCount, dimensions);
+        else {
+            queryVectors = TestUtils.generateRandomVectors(queryCount, dimensions);
+        }
+
+        float[][] groundTruthValues = TestUtils.computeGroundTruthDistances(indexVectors, queryVectors, spaceType, k);
+
+        List<List<float[]>> searchResults = new ArrayList<>();
+        List<float[]> kVectors;
+
+        for(int i = 0; i < queryVectors.length; i++){
+            KNNQueryBuilder knnQueryBuilderRecall = new KNNQueryBuilder(testField, queryVectors[i], k);
+            Response respRecall = searchKNNIndex(testIndex, knnQueryBuilderRecall,k);
+            List<KNNResult> resultsRecall = parseSearchResponse(EntityUtils.toString(respRecall.getEntity()), testField);
+
+            assertEquals(resultsRecall.size(), k);
+            kVectors = new ArrayList<>();
+            for(KNNResult result : resultsRecall){
+                kVectors.add(Floats.toArray(Arrays.stream(result.getVector()).collect(Collectors.toList())));
+            }
+            searchResults.add(kVectors);
+        }
+
+        return calculateRecallValue(queryVectors, searchResults, groundTruthValues, k, spaceType);
+    }
+
+    public double calculateRecallValue(float[][] queryVectors, List<List<float[]>> searchResults, float[][] groundTruthValues, int k, SpaceType spaceType){
+        ArrayList<Float> recalls = new ArrayList<>();
+        float dist = 0.0F;
+
+        for(int i = 0; i < queryVectors.length; i++){
+            float recallVal = 0.0F;
+            for(int j = 0; j < searchResults.get(i).size(); j++){
+                if(spaceType.getValue().equals("l2")){
```

Comment:
nit: please consider additional null check and order "constant equals variable"

Reply:
updated